### PR TITLE
RFC: Fix URTS

### PR DESCRIPTION
**Pull request by GalRogozinski**
[Rendered](https://github.com/GalRogozinski/protocol-rfcs/blob/urts/text/0008-uniform-random-tip-selection/0008-uniform-random-tip-selection.md)

Describes how current tip-selection works in Chrysallis

### CODE REVIEW

File: text/0008-uniform-random-tip-selection/0008-uniform-random-tip-selection.md [hunk 1]
```diff
@@ -1,206 +1,156 @@
...
+the cone it approves and previous issued  milestones.
+
+A tip can have one of 3 score states:
+* `0`: The tip is lazy and should not be selected.
+* `1`: The tip is somewhat lazy.
+* `2`: The tip is a non-lazy tip.
+
+Definitions:
+* `C1`: Max allowed delta value for the `YTRSI` of a given transaction in relation to the current `LSMI`.
+* `C2`: Max allowed delta value between `OTRSI` of the approvees of a given transaction in relation to the current `LSMI`.
```
Comment (karimodm) [inline text/0008-uniform-random-tip-selection/0008-uniform-random-tip-selection.md:63]:
```suggestion
* `C2`: Max allowed delta value between `OTRSI` of a given transaction in relation to the current `LSMI`. 
```
We don't calculate the OTRSI of the TX's approvees anymore, but of the TX itself.

File: text/0008-uniform-random-tip-selection/0008-uniform-random-tip-selection.md [hunk 2]
```diff
@@ -1,206 +1,156 @@
...
+in order to make it possible for it to be re-selected, which in turn makes the Tangle wider
+and improves synchronization speed. A tip is removed from the tips set if `X` amount of direct
+approvers are reached or if a certain amount of time `T` passed. 
+It is recommended to use `X` = 2 and `T` = 3  but the threshold should be configurable.
+
+### Purpose Of Semi-Lazy Tips
+
+Semi-Lazy tips are not eligible for tip-selection, but the coordinator node may implement a tip selection algorithm
+that confirms semi-lazy tips. Semi-lazy tips will usually be left behind, but parties interested in having them confirmed
+are incentivized to run spammers that will merge them into one thin subtangle.
```
Comment (karimodm) [inline text/0008-uniform-random-tip-selection/0008-uniform-random-tip-selection.md:123]: > will merge them into one thin subtangle. 

I find this sentence confusing: I would rephrase it in something like "... incentivized to run spammers that will actively reduce the amount of semi-lazy tips eligible for coordinator's tip selection"

File: text/0008-uniform-random-tip-selection/0008-uniform-random-tip-selection.md [hunk 3]
```diff
@@ -1,206 +1,156 @@
...
+
+# Rationale and alternatives
+
+The previous tip-selection was written in accordance to the original IOTA whitepaper, as it also
+functioned as the consensus mechanism to determine a transaction's confirmation rate.
+However, relatively soon it became apparent that the cumulative weight computation was too heavy
+for an actual high throughput scenario and as such, the CW calculation is currently not used within
+node implementations at all.
+
+Because confirmations with the `white-flag` approach no longer only approve cones with state mutations
```
Comment (karimodm) [inline text/0008-uniform-random-tip-selection/0008-uniform-random-tip-selection.md:142]: Link to WF for reference?

**Review by karimodm — CHANGES_REQUESTED** [review]
Thanks for taking care of fixing this man!

**Review by karimodm — APPROVED** [review]
Coolio ;)

**Review by Wollac — APPROVED** [review]

File: text/0008-uniform-random-tip-selection/0008-uniform-random-tip-selection.md [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+# Motivation
+
+Because of the `white-flag` confirmation algorithm, it is no longer necessary to perform complex
+tip-selection which evaluates ledger mutations while walking, therefore, a simpler and better 
+performing algorithm can be used to select tips, which in turn increases overall message throughput.
+
+In order to maximize confirmation rate however, the algorithm needs to return tips which are `non-lazy`.
+Non-lazy in this context means that a tip does not attach to a cone of messages which is too far
+in the past. Such a cone is likely to be already confirmed and does not contribute to the
+rate of newly confirmed messages when a milestone is issued.
```
Comment (charlesthompson3) [inline text/0008-uniform-random-tip-selection/0008-uniform-random-tip-selection.md:19]:
```suggestion
# Summary

Weighted Uniform Random Tip Selection on a subset enables a node to perform fast tip-selection to increase message throughput.
The algorithm selects tips which are non-lazy to maximize confirmation rate.

# Motivation

Because of the `white-flag` confirmation algorithm, it is no longer necessary to perform complex
tip-selection which evaluates ledger mutations while walking. Therefore, a more simple, better 
performing algorithm can be used to select tips, which in turn increases overall message throughput.

To maximize confirmation rate however, the algorithm needs to return tips which are `non-lazy`.
Non-lazy in this context means that a tip does not attach to a cone of messages which is too far
in the past. Such a cone is likely to be already confirmed and does not contribute to the
rate of newly confirmed messages when a milestone is issued.
```

File: text/0008-uniform-random-tip-selection/0008-uniform-random-tip-selection.md [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+* `Youngest Message Root Snapshot Index (YMRSI)` defines the highest milestone index of a set of
+`Confirmed Root Messages` of a given message.
+* `Latest Solid Milestone Index (LSMI)` the index of the latest solid milestone.
+* `Below Max Depth (BMD)` defines a threshold value up on which it is decided on whether a message is not
+relevant in relation to the recent parts of the Tangle. The current `BMD` for mainnet nodes is 15 milestones, 
+which means that messages of which their `OMRSI` in relation to the `LSMI` is more than 15, are "below max depth".
+
+### OMRSI / YMRSI example
+Given the blue PoV message, the `OMRSI` of it is milestone 1 and `YMRSI` milestone 2.
+Note that here again, the milestones are also `Confirmed Root Messages`.
```
Comment (charlesthompson3) [inline text/0008-uniform-random-tip-selection/0008-uniform-random-tip-selection.md:46]:
```suggestion
Note that, here again, the milestones are also `Confirmed Root Messages`.
```

File: text/0008-uniform-random-tip-selection/0008-uniform-random-tip-selection.md [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    }
+
+    return Score.NON_LAZY
+}
+```
+
+### Random Tip-Selection
+
+A node should keep a set of non-lazy tips (score 2).
+Every time a node is asked to select to tips to be approved it will just pick randomly from the set.
```
Comment (charlesthompson3) [inline text/0008-uniform-random-tip-selection/0008-uniform-random-tip-selection.md:108]:
```suggestion
Every time a node is asked to select tips to be approved, it will pick randomly from the set. 
```

File: text/0008-uniform-random-tip-selection/0008-uniform-random-tip-selection.md [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+The previous tip-selection was written in accordance to the original IOTA whitepaper, as it also
+functioned as part of the consensus mechanism.
+However, relatively soon it became apparent that the cumulative weight computation was too heavy
+for an actual high throughput scenario and as such, the CW calculation is currently not used within
+node implementations at all.
+
+Because confirmations with the [white-flag](https://github.com/iotaledger/protocol-rfcs/blob/master/text/0005-white-flag/0005-white-flag.md) approach no longer only approve cones with state mutations
+which are consistent with a previous ledger state, it makes sense to alter the tip-selection to provide 
+a fast way to get tips to approve with one's own message.
+The only important thing is to disincentive lazy behaviour in order to be able to maximize confirmation rate.
```
Comment (charlesthompson3) [inline text/0008-uniform-random-tip-selection/0008-uniform-random-tip-selection.md:143]:
```suggestion
A node must not execute tip-selection if it is not synchronized.

A tip should not be removed from the tips set immediately after it was selected in `select()`,  to make it possible for it to be re-selected, which in turn makes the Tangle wider
and improves synchronization speed. A tip is removed from the tips set if `X` amount of direct
approvers are reached or if a certain amount of time `T` passed. 
It is recommended to use `X` = 2 and `T` = 3  but the threshold should be configurable.

### Purpose Of Semi-Lazy Tips

Semi-Lazy tips are not eligible for tip-selection, but the coordinator node may implement a tip selection algorithm
that confirms semi-lazy tips. Semi-lazy tips will usually be left behind, but parties interested in having them confirmed
are incentivized to run spammers that will actively reduce the amount of semi-lazy tips eligible for coordinator's tip selection. 
Given a coordinator that chooses semi-lazy tips, running such spammers may get those messages confirmed before
they become lazy.


# Drawbacks

Depending on when and how often `YMRSI`/`OMRSI` values are computed, this tip-selection could still
have a slow runtime, as one would need to constantly walk down the Tangle to compute those
values. However, smart caching might resolve this issue. 

# Rationale and alternatives

The previous tip-selection was written in accordance to the original IOTA whitepaper, as it also
functioned as part of the consensus mechanism.
However, relatively soon it became apparent that the cumulative weight computation was too heavy
for an actual high throughput scenario and, as such, the CW calculation is currently not used within
node implementations at all.

Because confirmations with the [white-flag](https://github.com/iotaledger/protocol-rfcs/blob/master/text/0005-white-flag/0005-white-flag.md) approach no longer approve cones only with state mutations,
which are consistent with a previous ledger state, it makes sense to alter the tip-selection to provide 
a fast way to get tips to approve with one's own message.
The only important thing is to disincentive lazy behaviour to be able to maximize confirmation rate.
```

**Review by charlesthompson3 — COMMENTED** [review]
Here a few minor edits for clarity and grammar; please let me know if you have any questions / corrections to the suggestions!

-Charles

**Review by luca-moser — APPROVED** [review]
